### ISSUE TITLE
Trial test_getTimezoneOffset should be robust to platform specific behavior

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6846]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)| @oberstet reported|
|-|-|
|Trac ID|trac#6846|
|Type|enhancement|
|Created|2013-11-19 23:59:06Z|
|Branch|https://github.com/twisted/twisted/tree/test_log-utc-6846|

I'm not sure if the issue title is appropriate, nor if this is a "bug" in the Twisted test or in Python or one or more platforms.

The issue manifests on FreeBSD 9.2 in the test

```
twisted.test.test_log.FileObserverTestCase.test_getTimezoneOffset
```

This is currently the last test failing on that platform for select/poll reactors as can be seen in the log http://buildbot.twistedmatrix.com/builders/freebsd-9.2-amd64-python2.7/builds/5/steps/select/logs/problems.

Here is my analysis (also see the attached minimal reproducing code examples in the attachment).

 1. The UTC timezone has no daylight saving time (DST), since UTC is simply the reference zone, and by definition is fixed. See answer http://stackoverflow.com/a/5495816/884770.

 2. The Python standard library function `time.mktime` takes a 9-tuple where the last item is a flag `tm_isdst` that can take 3 values `0`, `1` or `-1`.

 3. The Python docs http://docs.python.org/2/library/time.html#time.mktime for that function say:

 "If the input value cannot be represented as a valid time, either OverflowError or ValueError will be raised (which depends on whether the invalid value is caught by Python or the underlying C libraries)."

 4. Now, if the host timezone is UTC, the behavior of `time.mktime` with `tm_isdst == 1` (that is, apply DST) is essentially undefined and platform specific. This can be seen from the attachment logs:

Windows also ignores the flag and applies the DST time difference originally in place when Python was started (in my case that was non-zero .. I did not want to change my TZ to check how it would cope with UTC .. and I don't know if that is possible on Windows anyway). IMO, this is clearly broken. UTC simply does not have DST, so the returned value is plain wrong.

Linux simply ignores the flag and assumes a DST time difference of 0. One could either say, ok, a non-existent DST is equivalent to a DST of 0, or could say, it's plain wrong, since non-existence isn't the same as zero.

FreeBSD recognizes the function argument as bogus (UTC timezone, but DST should be applied, when there is no DST for UTC) and raises an exception. IMO, this is the correct behavior. A program requesting an DST adjustment in the UTC timezone is doing it wrong, and should be signaled .. via exception.

Now, there is a flag `time.daylight` which the Python docs describe as "Nonzero if a DST timezone is defined.". This flag seems to signal a timezone that includes a DST, but the flag seems to be set on Python startup.  It changes if `os.environ['TZ']` is modified *and* `time.tzset()` is then called.

So far, so good;)

Now, who is responsible, where should it be fixed? The platforms, Python, Santa Claus?

I'd like to take a pragramtic approach .. and attach a proposed "fix".

Attachments:

* [fix_6846.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f47/f47df51ff80ec5044c45d5d466a9b5cdb230eb07/ec5233592bb3ca96010fe9a651015085d2fc4632.patch) (897 bytes) - added by oberstet on 2013-11-20 00:00:30Z - 
* [mktime_differences.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f47/f47df51ff80ec5044c45d5d466a9b5cdb230eb07/8a6b88b2e490f6020776a7f2e253c4cc7b9a5fbd.txt) (4024 bytes) - added by oberstet on 2013-11-20 00:02:30Z - minimal reproducing code
* [fix_6846.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f47/f47df51ff80ec5044c45d5d466a9b5cdb230eb07/1812cc9eb8314d5be8a50f0205fe91363575e7e6.patch) (2911 bytes) - added by oberstet on 2013-11-24 10:03:18Z - corrected and tested

<details><summary>Searchable metadata</summary>

```
trac-id__6846 6846
type__enhancement enhancement
reporter__oberstet oberstet
priority__normal normal
milestone__ 
branch__branches_test_log_utc_6846 branches/test_log-utc-6846
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1384905546000000 1384905546000000
changetime__1386444087000000 1386444087000000
version__None None
owner__lewq lewq
cc__oberstet
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Fixing a small inaccuracy in the handling of `time.daylight`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I'm confused by the attached patch.  It doesn't fix the issue.  It just changes the problem to a `NameError` when TZ=UTC.

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_3"></a>|@oberstet commented|
|-|-|

How do you trigger the `NameError`? I've tried hard, please see below. Both systems are native UTC timezone.

**Linux**

```
[ec2-user@ip-10-58-91-252 ~]$ python
Python 2.6.8 (unknown, Mar 14 2013, 09:31:22)
[GCC 4.6.2 20111027 (Red Hat 4.6.2-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> time.daylight
0
>>> import os
>>> os.environ['TZ']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.6/UserDict.py", line 22, in __getitem__
    raise KeyError(key)
KeyError: 'TZ'
>>>
[ec2-user@ip-10-58-91-252 ~]$ date
Thu Nov 21 21:34:07 UTC 2013
[ec2-user@ip-10-58-91-252 ~]$ cat /etc/localtime
TZif2UTCTZif2UTC
UTC0
[ec2-user@ip-10-58-91-252 ~]$ export TZ='UTC'
[ec2-user@ip-10-58-91-252 ~]$ python
Python 2.6.8 (unknown, Mar 14 2013, 09:31:22)
[GCC 4.6.2 20111027 (Red Hat 4.6.2-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> time.daylight
0
>>> import os
>>> os.environ['TZ']
'UTC'
>>>
```


**FreeBSD**

```
[oberstet@tvd_build_txpypy ~]$ python
Python 2.7.6 (default, Nov 20 2013, 09:08:10)
[GCC 4.2.1 Compatible FreeBSD Clang 3.3 (tags/RELEASE_33/final 183502)] on freebsd9
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> time.daylight
0
>>>
[oberstet@tvd_build_txpypy ~]$ cat /etc/localtime
TZif2UTCTZif2UTC
UTC0
[oberstet@tvd_build_txpypy ~]$ export TIMEZONE='UTC'
[oberstet@tvd_build_txpypy ~]$ export TZ='UTC'
[oberstet@tvd_build_txpypy ~]$ python
Python 2.7.6 (default, Nov 20 2013, 09:08:10)
[GCC 4.2.1 Compatible FreeBSD Clang 3.3 (tags/RELEASE_33/final 183502)] on freebsd9
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> time.daylight
0
>>>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

```
exarkun@top:~/Projects/Twisted/trunk$ TZ=UTC bin/trial twisted.test.test_log.FileObserverTestCase.test_getTimezoneOffset
twisted.test.test_log
  FileObserverTestCase
    test_getTimezoneOffset ...                                          [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_log.py", line 383, in test_getTimezoneOffset
    self.flo.getTimezoneOffset(utcDaylightTimestamp),
exceptions.UnboundLocalError: local variable 'utcDaylightTimestamp' referenced before assignment

twisted.test.test_log.FileObserverTestCase.test_getTimezoneOffset
-------------------------------------------------------------------------------
Ran 1 tests in 0.041s

FAILED (errors=1)
exarkun@top:~/Projects/Twisted/trunk$ 
```

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)|@oberstet set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Thanks for working on this oberstet.  I don't think the change here preserves the intent of the test though.  It just disables it when the timezone inherited from the environment has certain properties (specifically, not having DST).

This dependence is a bug in the test (one that I'm quite surprised no one ever noticed before since it shows up whenever TZ=UTC, but oh well).  It fails to construct a POSIX timestamp to use as test data when TZ=UTC but the actual behavior it is testing is all for cases where TZ != UTC.  So instead I suggest that the test just shouldn't call `mktime` with an unknown TZ.  This can easily be done by not calling `mktime` until the timezone has been changed (which the test does three times).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [40860]) Branching to 'test_log-utc-6846'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(http://bugs.python.org/issue15750 may be relevant.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Based on the info in the python bug report, I concluded that only UTC presents this problem.  Therefore I did not try to address the fact that mktime with a DST time tuple in the Africa/Johannesburg timezone doesn't make sense (because it works anyway).

Additional build links (for not-yet-supported platforms where this fix is particularly relevant:

 * https://buildbot.twistedmatrix.com/builders/freebsd-9.2-amd64-pypy2.2/builds/10
 * https://buildbot.twistedmatrix.com/builders/freebsd-9.2-amd64-pypy-nightly/builds/8
 * https://buildbot.twistedmatrix.com/builders/freebsd-9.2-amd64-python2.7/builds/15

The test appears to pass there now.  They pass on supported platforms too, it seems (with spurious errors reported from the pyflakes and winxp32-py2.7 builders).

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_11"></a>|@oberstet commented|
|-|-|

Awesome! Much better to not entirely skip, but create a well-defined test setup by setting TZs. To me, the code looks fine.

Regarding your comments (above and in the patch) Africa/Johannesburg not having a DST. In my understanding, the difference to UTC is as follows.

UTC does not have a DST by definition. Never, neither in past, nor presence, _nor_ in any future, since it's by definition.

Africa/Johannesburg does not have a DST as of today (2013). However, it did have one in the past: Johannesburg's last DST ended on Sun 19-Mar-1944 at 02:00:00 A.M. when local clocks were set backward 1 hour. And in fact it might have a DST in the future again. I guess that is a political decision. We don't know. But for UTC we know: it won't ever have a DST.

So an implementation might choose to interpret a request for adjusting to DST for a timezone that _currently_ does not have a DST by just applying a delta of 0 hours (which seems to be Linux behavior). But it might choose to raise an error when a request is done for adjusting to DST for the UTC "timezone" (and only that) - since it just doesn't make sense (which seems to be FreeBSD behavior). A third implementation might choose to raise either when DST adjustment is requested for UTC or for a TZ that _currently_ does not have a DST (which platform I don't know).

I might be wrong with above argumentation - time is an incredible complex concept.

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_12"></a>|@oberstet commented|
|-|-|

That last paragraph of my last comment - correction:

"So an implementation might choose to interpret a request for adjusting to DST both a timezone that _currently_ does not have a DST _and_ for the UTC "timezone" by just applying a delta of 0 hours (which seems to be Linux behavior). Or/But .."

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lewq set owner to lewq|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|lewq commented|
|-|-|

Thanks for this. The first issue I have with the patch is that while the comments above talk about UTC being a special case, UTC itself is not tested in the code. Can you address this and resubmit, or explain why that's not necessary?

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_17"></a>|@oberstet commented|
|-|-|

The special case is that the _host_ on which trial runs has it's timezone set to UTC. The trial cannot change the host's timezone.

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|lewq commented|
|-|-|

So do we have buildslaves with UTC and non-UTC timezones set, in order to get coverage on all the cases?

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_19"></a>|@oberstet commented|
|-|-|

Since the behavior of the relevant OS API differs between Windows, Linux and FreeBSD (and possibly OSX), you are now asking if we have buildslave for all these systems one with UTC timezone and one with non-UTC each? I guess the answer is no. It's a mixture.

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|lewq commented|
|-|-|

Please add documentation on the magic times specified to convince me that they are/are not actually daylight saving in both localities (New York and Berlin).  Please also test the UTC case, then merge.  Thanks!

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|lewq commented|
|-|-|

> The special case is that the _host_ on which trial runs has it's timezone set to UTC. The trial cannot change the host's timezone.

I think we are fairly confident that trial *can* effectively set the timezone of the host, since the updated test does exactly that. But I may have missed the point of the implications of this comment.

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_22"></a>|@oberstet commented|
|-|-|

The test doesn't set the timezone on the host, but only modifies the trial process environment variable TZ. At least on our buildslave, the Unix user under which trial runs has no write access to `/etc`.

The UTC case cannot be added to the test, since the tests are calling mktime with tm_isdst == 1 to check daylight saving adjustment, which will fail for TZ == UTC on certain platforms, which is the whole point of the issue at hand.

Regarding documentation: you want me to document the fact that NY and Berlin do not have the _same_ daylight saving?

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|lewq commented|
|-|-|

I think I understand this issue better now.  The point of this ticket is that the test in question should be invariant under this platform-specific behaviour (which is what the change in [#6846](https://github.com/twisted/twisted/issues/6846)#comment:10 achieves).  Documenting (in the form of additional platform-specific tests) the different platform-specific behaviours might be nice, in order to try to get more coverage, but perhaps this is outside the scope of this ticket.  What do you think, oberstet/exarkun?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [40950]) Yup, that does not work.

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/bbslave_twisted/work/freebsd-9.2-amd64/Twisted/twisted/test/test_log.py", line 457, in test_getTimezoneOffsetAtUTC
    self._getTimezoneOffsetTest("UTC", 0, 0)
  File "/home/bbslave_twisted/work/freebsd-9.2-amd64/Twisted/twisted/test/test_log.py", line 395, in _getTimezoneOffsetTest
    daylight = time.mktime(localDaylightTuple)
exceptions.OverflowError: mktime argument out of range

twisted.test.test_log.FileObserverTestCase.test_getTimezoneOffsetAtUTC
-------------------------------------------------------------------------------
```


refs [#6846](https://github.com/twisted/twisted/issues/6846)

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_25"></a>|@oberstet commented|
|-|-|

@lewq:

Regarding your first sentance: yes, exactly.

Regarding the 2nd point: this would effectively mean writing tests against the Posix API mktime() and/or the Python wrapper around this API. I'm not even sure if the relevant Posix standard has a well defined spec for mktime() (that clearly and uniquely defines the desired behavior). And even if so: to whom will we file bugs when the test fails? Python? Microsoft? I can't see how this should be Twisted's business - but I might be wrong: I lack knowledge about the Twisted project policy in this respect. But in any case, I do think this would then be better a separate ticket (like: test Posix mktime() implementation). But really, I think this is getting far off ..

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

Latest version of the branch tries to clarify the "daylight" vs "standard" timestamps being constructed.  It also splits the one big test into three little tests (which is a bit of unasked-for refactoring but it let me run the UTC test explicitly and independently on buildbot and get the above copied result showing it isn't workable there, then I deleted the test).

So I think this is probably good to merge now.  If latest build turns up green then I will.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [40953]) Merge test_log-utc-6846 

Author: oberstet, exarkun
Reviewer: exarkun, oberstet, lewq
Fixes: [#6846](https://github.com/twisted/twisted/issues/6846)

Avoid calling mktime() with invalid arguments in the UTC timezone
since this provokes an error from certain platforms.
